### PR TITLE
refactor: new api tokens

### PR DESCRIPTION
**Pull request by agualis**
Introduces a new `ApiToken` TS type with the attributes shared by  `GqlToken` and `pool.poolTokens`.

**How to test**: 
Temporary comment  pool '`0xc832a37c8252117604f1329b4a7fed7076880b27`' from sepolia `allowNestedActions` here:
https://github.com/balancer/frontend-monorepo/blob/d3e3fdb22b7dd18dc31905dd317f87f0debcbef2/packages/lib/config/networks/sepolia.ts#L55

Then you can add using the nested BPT token: 

Screenshot 2024-12-13 at 14 29 30

Liquidity Pool (v3): Balancer DAI:USDC:USDT 2024-12-13 14-29-50

### CODE REVIEW

File: packages/lib/modules/pool/actions/add-liquidity/form/useMaximumInputs.tsx [hunk 1]
```diff
@@ -40,6 +40,7 @@ export function useMaximumInputs() {
       .map(balance => ({
         humanAmount: balance.formatted as HumanAmount,
         tokenAddress: balance.address as Address,
+        symbol: 'TBD', // TODO: finish with Groninge
```
Comment (garethfuller) [inline packages/lib/modules/pool/actions/add-liquidity/form/useMaximumInputs.tsx:43]: If this isn't changed it blocks the PR being merged?

File: packages/lib/modules/pool/actions/unstake/UnstakeProvider.tsx [hunk 2]
```diff
@@ -35,14 +35,13 @@ export function _useUnstake() {
...
-        tokenAddress: reward.tokenAddress,
-        humanAmount: reward.humanBalance as HumanAmount,
-      })),
-    [allClaimableRewards]
-  )
+  const rewardAmounts: HumanTokenAmount[] = useMemo(() => {
+    return allClaimableRewards.map(reward => ({
+      tokenAddress: reward.tokenAddress,
+      humanAmount: reward.humanBalance as HumanAmount,
+      symbol: 'LP token',
```
Comment (garethfuller) [inline packages/lib/modules/pool/actions/unstake/UnstakeProvider.tsx:42]: This is wrong. Reward amounts are not LP tokens. You need the actual symbol here.

File: packages/lib/modules/pool/actions/LiquidityActionHelpers.ts [hunk 3]
```diff
@@ -412,12 +427,14 @@ export function formatBuildCallParams<T>(buildCallParams: T, account: Address) {
...
 export function toTokenAmountsIn(
-  sdkQueryOutput: AddLiquidityQueryOutput
+  sdkQueryOutput: AddLiquidityQueryOutput,
+  pool: Pool
 ): TokenAmountIn[] | undefined {
   if (!sdkQueryOutput) return
   return sdkQueryOutput.amountsIn.map(amountIn => ({
     address: amountIn.token.address,
     amount: amountIn.amount,
+    symbol: amountIn.token.symbol || getActionableTokenSymbol(amountIn.token.address, pool),
```
Comment (garethfuller) [inline packages/lib/modules/pool/actions/LiquidityActionHelpers.ts:437]: Why not just a more generic function which searches all the pool's token for a
matching token to get the symbol? e.g. `getPoolTokenSymbol(address, pool)`

File: packages/lib/modules/pool/pool.helpers.ts [hunk 4]
```diff
@@ -420,33 +421,43 @@ export function getWarnings(warnings: string[]) {
...
+  const token = getPoolActionableTokens(pool).find(token =>
+    isSameAddress(token.address, tokenAddress)
+  )
+  if (!token) {
+    console.log('Token symbol not found for address ', tokenAddress)
+    return ''
+  }
+
+  return token.symbol
+}
```
Comment (garethfuller) [inline packages/lib/modules/pool/pool.helpers.ts:434]: Same comment, make it more generic, doesn't need to be limited to the action tokens.

File: packages/lib/modules/portfolio/PortfolioClaim/ClaimProtocolRevenueModal.tsx [hunk 5]
```diff
@@ -51,11 +51,12 @@ export default function ClaimProtocolRevenueModal({ isOpen, onClose }: Props) {
 
   const claimTxHash = transactionSteps.lastTransaction?.result?.data?.transactionHash
 
-  const rewards: HumanTokenAmountWithAddress[] = rewardsDataSnapshot
+  const rewards: HumanTokenAmount[] = rewardsDataSnapshot
     .filter(reward => !bn(reward.balance).isZero())
     .map(reward => ({
       tokenAddress: reward.tokenAddress as Address,
       humanAmount: reward.humanBalance,
+      symbol: '-',
```
Comment (garethfuller) [inline packages/lib/modules/portfolio/PortfolioClaim/ClaimProtocolRevenueModal.tsx:59]: Again, why do this? Will it not cause problems or bad UX?

File: packages/lib/modules/swap/usePermit2SwapStep.tsx [hunk 6]
```diff
@@ -48,6 +48,7 @@ export function useSignPermit2SwapStep({
   const tokenIn: TokenAmountIn = {
     address: tokenInAddress,
     amount: getTokenInAmount(),
+    symbol: tokenInInfo?.symbol ?? '',
```
Comment (garethfuller) [inline packages/lib/modules/swap/usePermit2SwapStep.tsx:51]: If it's not required can we not make it optional on the type?

File: packages/lib/modules/tokens/token.helpers.ts [hunk 7]
```diff
@@ -136,14 +136,24 @@ export function getLeafTokens(poolTokens: PoolToken[]) {
...
+function getTokenOrUnderlying(token: PoolToken): ApiToken {
   return token.isErc4626 && token.underlyingToken
     ? {
         ...token.underlyingToken,
         address: token.underlyingToken.address as Address,
-        index: token.index,
+        chainId: token.underlyingToken.chainId as number,
+        chain: token.underlyingToken.chain as GqlChain,
+        priority: token.underlyingToken.priority as number,
+        tradable: token.underlyingToken.tradable as boolean,
```
Comment (garethfuller) [inline packages/lib/modules/tokens/token.helpers.ts:147]: Create a `buildApiToken` for doing this?

File: packages/lib/modules/pool/actions/unstake/UnstakeProvider.tsx (hunk 2 above)
Comment (agualis) [reply packages/lib/modules/pool/actions/unstake/UnstakeProvider.tsx:42]: You are right. I had tested this and it was working as expected cause we don't really need the symbol here (I added it by mistake during the refactor) but we are getting it with `getToken` as we don't have the reward token symbol in the rewards onchain queries. 

Examples after removing: 

Screenshot 2024-12-17 at 17 31 33

Screenshot 2024-12-17 at 17 33 17

File: packages/lib/modules/tokens/token.helpers.ts (hunk 7 above)
Comment (agualis) [reply packages/lib/modules/tokens/token.helpers.ts:147]: Extracted `buildApiToken` and added a TODO that I will address with you in slack.

File: packages/lib/modules/pool/actions/LiquidityActionHelpers.ts (hunk 3 above)
Comment (agualis) [reply packages/lib/modules/pool/actions/LiquidityActionHelpers.ts:452]: I prefer to wait until we do the next refactor that clearly splits between 3 sets of tokens (Header/composition/liquidity actions). (added a TODO and ticket for that):
https://github.com/balancer/frontend-monorepo/blob/a898f5ab1b14c951b11cf394cf9476306249d6e3/packages/lib/modules/pool/pool.helpers.ts#L424

File: packages/lib/modules/portfolio/PortfolioClaim/ClaimProtocolRevenueModal.tsx (hunk 5 above)
Comment (agualis) [reply packages/lib/modules/portfolio/PortfolioClaim/ClaimProtocolRevenueModal.tsx]: symbol was not needed here. I added it during an alternative refactor and forgot about removing it later.

File: packages/lib/modules/pool/actions/add-liquidity/form/useMaximumInputs.tsx (hunk 1 above)
Comment (agualis) [reply packages/lib/modules/pool/actions/add-liquidity/form/useMaximumInputs.tsx:43]: this is the last missing issue. I will try to figure it out without Groninge tomorrow (he's quite busy)

Comment (agualis) [reply packages/lib/modules/pool/actions/add-liquidity/form/useMaximumInputs.tsx]: Removed `useMaximumInputs` as it is not used anymore.

File: packages/lib/modules/pool/pool.helpers.ts [hunk 8]
```diff
@@ -29,8 +29,9 @@ import { supportsNestedActions } from './actions/LiquidityActionHelpers'
 import { getLeafTokens, PoolToken } from '../tokens/token.helpers'
 import { GetTokenFn } from '../tokens/TokensProvider'
 import { vaultV3Abi } from '@balancer/sdk'
-import { TokenCore, PoolListItem } from './pool.types'
+import { TokenCore, PoolListItem, ApiToken } from './pool.types'
 import { Pool } from './PoolProvider'
+import { HumanTokenAmountWithAddress } from '../tokens/token.types'
```
Comment (garethfuller) [inline packages/lib/modules/pool/pool.helpers.ts:34]: Remove unused import

File: packages/lib/modules/tokens/token.helpers.ts [hunk 9]
```diff
@@ -11,7 +11,7 @@ import { HumanTokenAmountWithAddress, TokenBase } from './token.types'
 import { InputAmount } from '@balancer/sdk'
 import { Pool } from '../pool/PoolProvider'
 import { getVaultConfig, isCowAmmPool, isV3Pool } from '../pool/pool.helpers'
-import { TokenCore } from '../pool/pool.types'
+import { ApiToken, TokenCore } from '../pool/pool.types'
```
Comment (garethfuller) [inline packages/lib/modules/tokens/token.helpers.ts:14]: Remove unused import

**Review by garethfuller — APPROVED** [review]

**Pull request merged**